### PR TITLE
Archival

### PR DESCRIPTION
**Pull request by lukasjuhrich**
Prepares #490 
- [x] Discuss whether we want to do this yearly or regularly && adapt code accordingly
- [x] Test archival selection
  - [x] test `select_user_and_last_mem()` select (used as CTE)
  - [x] test cutoff filtering
  - [x] test cutoff calculation from current date
  - [x] test `do-not-archive` filtering
- [ ] ~~Implement `archive_users` (cleanup)~~
- [ ] ~~Add a web endpoint and button archiving all current users~~
- [ ] ~~Add a celerybeat task executing the archival~~

_(update 2026-02-01: in the interest of consolidating changes and moving this forward, we will tackle the follow-up items in separate PRs)_

### CODE REVIEW

**Comment by lukasjuhrich** [comment]
This PR, as of now, does some base work that can directly be merged into `develop`.
All the rest can be coordinated in one or multiple different PRs (see the extended tasklist of #67).
Waiting for the tests to finish.

File: pycroft/lib/user_deletion.py [hunk 1]
```diff
@@ -65,20 +42,59 @@ def get_archivable_members(session: Session) -> Sequence[ArchivableMemberInfo]:
...
+def get_archivable_members(
+    session: Session,
+    current_year: int | None = None,
+) -> Sequence[ArchivableMemberInfo]:
+    """Return all the users that qualify for being archived right now.
 
-def get_invalidated_archive_memberships() -> list[Membership]:
-    """Get all memberships in `to_be_archived` of users who have an active `do-not-archive` property.
+    Selected are those users
+    - whose last membership in the member_group ended two weeks in the past,
```
Comment (georgeto) [inline pycroft/lib/user_deletion.py:62]: Looks do me like we are now testing for >= 2 years and not weeks.

Comment (lukasjuhrich) [reply pycroft/lib/user_deletion.py]: Fixed in ac859d266f0d

File: pycroft/lib/user_deletion.py [hunk 2]
```diff
@@ -76,20 +67,278 @@ def get_archivable_members(session: Session) -> Sequence[ArchivableMemberInfo]:
...
+    We joined load the following information:
+    - hosts
+    - account
+    - room
+    - current_properties_maybe_denied
+
+    :param session:
+    :param current_year: dependency injection of the current year.
+    :param years_following_eom: dependency injection of the current year.
+        defaults to the current year.
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply pycroft/lib/user_deletion.py]: fixed

File: tests/lib/scrubbing/conftest.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+@pytest.fixture(scope="module", autouse=True)
+def users_do_not_archive(
+    module_session: Session, config: Config, group_do_not_archive: PropertyGroup
+) -> list[User]:
+    """
+    Create a user with a membership in a group that has the do-not-archive
+    property.
+    """
+    # TODO wayyy better example set! instead of making all `do-not-archive`,
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply tests/lib/scrubbing/conftest.py:54]: wontfix ;)

File: web/blueprints/user/__init__.py [hunk 4]
```diff
@@ -1492,8 +1500,30 @@ def resend_confirmation_mail() -> ResponseReturnValue:
...
+def scrub_mails() -> ResponseReturnValue:
+    """Scrub The mail addresses in batches of 100"""
+    sess = t.cast(Session, session.session)
+
+    with sess.begin_nested():
+        ids = scrub_all_mails(sess, author=current_user)
+
+    sess.commit()
+
+    flash(f"{len(ids.all())} E-Mail-Adressen gelöscht.", "success")
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply web/blueprints/user/__init__.py:1524]: I'm confused by this: since when are `ScalarResult`s materialized? they still have `fetchmany`, `all` as methods, and  for good reason. The latter calls `ResultInternal._allrows()`.

File: web/blueprints/user/__init__.py [hunk 5]
```diff
@@ -1492,8 +1500,30 @@ def resend_confirmation_mail() -> ResponseReturnValue:
...
-    return render_template('user/archivable_users.html', table=table)
+    table = ArchivableMembersTable(data_url=url_for(".archivable_users_json"))
+
+    sess = t.cast(Session, session.session)
+
+    return render_template(
+        "user/archivable_users.html",
+        table=table,
+        archivable_mail_count=scrubbable_mails_count(sess, datetime.now().year),
+        archivable_mail_users=next(scrubbable_mails(sess).partitions(10), []),
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply web/blueprints/user/__init__.py]: fixd

File: web/blueprints/user/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply web/blueprints/user/__init__.py:1524]: wontfix.

File: tests/lib/scrubbing/conftest.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+
+@pytest.fixture(autouse=True)
+def freeze(freezer) -> None:
+    freezer.move_to("2025-01-01")
+
+
+@pytest.fixture(scope="module")
+def user_archivable(module_session: Session, config: Config) -> User:
+    # TODO: user wtih membership
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply tests/lib/scrubbing/conftest.py]: removed

File: tests/lib/scrubbing/conftest.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        with_membership=True,
+        membership__active_during=closedopen(datetime(2020, 7, 1), datetime(2021, 11, 25)),
+        membership__group=config.member_group,
+        without_room=True,
+    )
+
+
+@pytest.fixture(scope="module")
+def users_archivable(module_session: Session, config: Config) -> list[User]:
+    # TODO: user wtih membership
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply tests/lib/scrubbing/conftest.py]: removed

File: pycroft/lib/membership.py [hunk 8]
```diff
@@ -239,3 +243,46 @@ def change_membership_active_during(
         .to_json()
     )
     log_user_event(message, processor, membership.user)
+
+
+def select_user_and_last_mem() -> Select[tuple[int, int, str]]:
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply pycroft/lib/membership.py]: fixed

File: pycroft/lib/user_deletion.py [hunk 9]
```diff
@@ -76,20 +67,274 @@ def get_archivable_members(session: Session) -> Sequence[ArchivableMemberInfo]:
...
+    session.add(le)
+    session.add(ScrubLog(scrubber="mail", info={"user_id": user.id}))
+
+
+def scrub_all_mails(
+    session: Session,
+    author: User,
+    user_filter: t.Callable[[Select[tuple[User]]], Select[tuple[User]]] | None = None,
+) -> ScalarResult[int]:
+    return session.scalars(scrub_all_mails_stmt(utcnow().year, author.id, user_filter))
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply pycroft/lib/user_deletion.py:167]: wrong, this is our `utcnow()` helper.

File: pycroft/lib/user_deletion.py [hunk 10]
```diff
@@ -76,20 +67,278 @@ def get_archivable_members(session: Session) -> Sequence[ArchivableMemberInfo]:
...
+                UserLogEntry.user_id.name,
+            ),
+            select(
+                logs_and_user_id.c.id,
+                logs_and_user_id.c.user_id,
+            ),
+        )
+        .cte("user_log_entries")
+    )
+    return select(removed_mail_user_ids.c.id).add_cte(insert_scrub_log, insert_user_log)
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply pycroft/lib/user_deletion.py]: AFAIK this is not true in postgres, and tests also indicate that they are executed properly.

File: tests/lib/scrubbing/conftest.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+def users_do_not_archive(
+    module_session: Session, config: Config, group_do_not_archive: PropertyGroup
+) -> list[User]:
+    """
+    Create a user with a membership in a group that has the do-not-archive
+    property.
+    """
+    # TODO wayyy better example set! instead of making all `do-not-archive`,
+    # we would want them to have different membership ranges, also additional memberships;
+    # also edge-cases like upcoming memberships or on-new-years membership terminations
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply tests/lib/scrubbing/conftest.py:56]: Sure, but that is not a blocker for me.

File: web/blueprints/user/__init__.py [hunk 12]
```diff
@@ -9,9 +9,10 @@
 
     :copyright: (c) 2012 by AG DSN.
 """
+from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (lukasjuhrich) [reply web/blueprints/user/__init__.py]: removed.
